### DISCUSSION TITLE
What's a good program to generate to showcase MIR?

### DISCUSSION THREAD

**Discussion by iacore** [vnmakarov/mir#381] [Q&A]
Hello @vnmakarov,

I want to write an article (blog post) showcasing MIR. However, I don't have a good idea of the program I would generate in MIR.

My best thought so far is to write a BrainF*ck interpreter/compiler. It is still a bit boring to me, since the BF idea is done by so many people.

Do you have a more novel idea that's also as simple as a BF interpreter?

Best Regards
@iacore

**Comment by vnmakarov** [comment]
Not really.  I thought about designing and implementing a simple dynamic programming language when all metaprogramming is implemented (https://developers.redhat.com/articles/2022/02/16/code-specialization-mir-lightweight-jit-compiler).  I thought about implementing only interpreter, annotate it and automatically generate JIT from the interpreter for it.  It is for MIR metaprogramming demonstration.

But it is not implemented yet, so I did not start to work on this.

In your case, I only would recommend to implement a simple **static type** programming language.  For static language, it is easy to demonstrate how performance JIT implemented by the current state MIR is superior to the interpreter one.
